File: framework/languages/java/e2e-kotlin/RegistryMessaging/Client/src/main/kotlin/systems/zlink/e2e/kotlin/registrymessaging/client/Support/ClientOptions.kt
package systems.zlink.e2e.kotlin.registrymessaging.client.Support

data class ClientOptions(
    val registryUrl: String,
    val providerAUrl: String,
    val providerBUrl: String,
    val workflowUrl: String,
    val directConsumerUrl: String,
    val singleConsumerUrl: String,
    val discoveryConsumerUrl: String,
    val backpressureConsumerUrl: String,
    val registryBin: String,
    val providerBin: String,
    val logDir: String,
    val scenario: String,
) {
    companion object {
        fun parse(args: Array<String>): ClientOptions {
            val values = linkedMapOf<String, String>()
            var index = 0
            while (index < args.size) {
                val key = args[index]
                if (!key.startsWith("--")) {
                    index++
                    continue
                }
                require(index + 1 < args.size) { "Missing value for $key." }
                values[key.removePrefix("--")] = args[index + 1]
                index += 2
            }
            fun required(key: String): String =
                values[key]?.takeIf { it.isNotBlank() } ?: throw IllegalArgumentException("--$key is required.")
            return ClientOptions(
                registryUrl = required("registry-url"),
                providerAUrl = required("provider-a-url"),
                providerBUrl = required("provider-b-url"),
                workflowUrl = required("workflow-url"),
                directConsumerUrl = required("direct-consumer-url"),
                singleConsumerUrl = required("single-consumer-url"),
                discoveryConsumerUrl = required("discovery-consumer-url"),
                backpressureConsumerUrl = required("backpressure-consumer-url"),
                registryBin = required("registry-bin"),
                providerBin = required("provider-bin"),
                logDir = required("log-dir"),
                scenario = values["scenario"]?.takeIf { it.isNotBlank() } ?: "all",
            )
        }
    }
}

File: framework/languages/dotnet/e2e/LocationMessaging/Client/Support/ClientOptions.cs
namespace LocationMessaging.Client.Support;

internal sealed record ClientOptions(
    string ProviderAUrl,
    string ProviderBUrl,
    string WorkflowUrl,
    string DirectConsumerUrl,
    string SingleConsumerUrl,
    string StoreConsumerUrl,
    string BackpressureConsumerUrl,
    string RedisEndpoint,
    string RedisKeyPrefix,
    string ProviderProject,
    string LogDir,
    string Scenario)
{
    public static ClientOptions Parse(string[] args)
    {
        var values = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        for (var i = 0; i < args.Length; i++)
        {
            var key = args[i];
            if (!key.StartsWith("--", StringComparison.Ordinal)) continue;

            if (i + 1 >= args.Length) throw new ArgumentException($"Missing value for {key}.");

            values[key[2..]] = args[++i];
        }

        return new ClientOptions(
            values.TryGetValue("provider-a-url", out var providerAUrl) && !string.IsNullOrWhiteSpace(providerAUrl)
                ? providerAUrl
                : throw new ArgumentException("--provider-a-url is required."),
            values.TryGetValue("provider-b-url", out var providerBUrl) && !string.IsNullOrWhiteSpace(providerBUrl)
                ? providerBUrl
                : throw new ArgumentException("--provider-b-url is required."),
            values.TryGetValue("workflow-url", out var workflowUrl) && !string.IsNullOrWhiteSpace(workflowUrl)
                ? workflowUrl
                : throw new ArgumentException("--workflow-url is required."),
            values.TryGetValue("direct-consumer-url", out var directConsumerUrl)
            && !string.IsNullOrWhiteSpace(directConsumerUrl)
                ? directConsumerUrl
                : throw new ArgumentException("--direct-consumer-url is required."),
            values.TryGetValue("single-consumer-url", out var singleConsumerUrl)
            && !string.IsNullOrWhiteSpace(singleConsumerUrl)
                ? singleConsumerUrl
                : throw new ArgumentException("--single-consumer-url is required."),
            values.TryGetValue("store-consumer-url", out var storeConsumerUrl)
            && !string.IsNullOrWhiteSpace(storeConsumerUrl)
                ? storeConsumerUrl
                : throw new ArgumentException("--store-consumer-url is required."),
            values.TryGetValue("backpressure-consumer-url", out var backpressureConsumerUrl)
            && !string.IsNullOrWhiteSpace(backpressureConsumerUrl)
                ? backpressureConsumerUrl
                : throw new ArgumentException("--backpressure-consumer-url is required."),
            values.TryGetValue("redis-endpoint", out var redisEndpoint)
            && !string.IsNullOrWhiteSpace(redisEndpoint)
                ? redisEndpoint
                : throw new ArgumentException("--redis-endpoint is required."),
            values.TryGetValue("redis-key-prefix", out var redisKeyPrefix)
            && !string.IsNullOrWhiteSpace(redisKeyPrefix)
                ? redisKeyPrefix
                : throw new ArgumentException("--redis-key-prefix is required."),
            values.TryGetValue("provider-project", out var providerProject)
            && !string.IsNullOrWhiteSpace(providerProject)
                ? providerProject
                : throw new ArgumentException("--provider-project is required."),
            values.TryGetValue("log-dir", out var logDir) && !string.IsNullOrWhiteSpace(logDir)
                ? logDir
                : throw new ArgumentException("--log-dir is required."),
            values.TryGetValue("scenario", out var scenario) && !string.IsNullOrWhiteSpace(scenario)
                ? scenario
                : "all");
    }
}
